lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py

```diff
@@ -247,92 +252,118 @@ def argspec(self):
-                if running:
-                    if obj.metadata.name == name and hasattr(obj, 'status'):
-                        phase = getattr(obj.status, 'phase', None)
-                        if phase:
-                            if phase == 'Running' and running:
-                                watcher.stop()
-                                return
-                else:
-                    # TODO: wait for stopped state:
-                    watcher.stop()
-                    return
-
-        self.fail_json(msg="Error waiting for virtual machine. Try a higher wait_timeout value. %s" % obj.to_dict())
-
-    def manage_state(self, state):
-        wait = self.params.get('wait')
-        wait_timeout = self.params.get('wait_timeout')
-        resource_version = self.params.get('resource_version')
-
-        resource_vm = self.find_supported_resource('VirtualMachine')
-        existing = self.get_resource(resource_vm)
-        if resource_version and resource_version != existing.metadata.resourceVersion:
-            return False
-
-        existing_running = False
-        resource_vmi = self.find_supported_resource('VirtualMachineInstance')
-        existing_running_vmi = self.get_resource(resource_vmi)
-        if existing_running_vmi and hasattr(existing_running_vmi.status, 'phase'):
-            existing_running = existing_running_vmi.status.phase == 'Running'
-
-        if state == 'running':
-            if existing_running:
-                return False
+    @staticmethod
+    def fix_serialization(obj):
+        if obj and hasattr(obj, 'to_dict'):
+            return obj.to_dict()
+        return obj
+
+    def _wait_for_vmi_running(self):
+        for event in self._kind_resource.watch(namespace=self.namespace, timeout=self.params.get('wait_timeout')):
+            entity = event['object']
+            if entity.metadata.name != self.name:
+                continue
+            status = entity.get('status', {})
+            phase = status.get('phase', None)
+            if phase == 'Running':
+                return entity
+
+        self.fail("Error waiting for virtual machine. Maybe try a higher wait_timeout value?")
+
+    def _wait_for_vm_state(self, new_state):
+        if new_state == 'running':
+            want_created = want_ready = True
+        else:
+            want_created = want_ready = False
+
+        for event in self._kind_resource.watch(namespace=self.namespace, timeout=self.params.get('wait_timeout')):
+            entity = event['object']
+            if entity.metadata.name != self.name:
+                continue
+            status = entity.get('status', {})
+            created = status.get('created', False)
+            ready = status.get('ready', False)
+            if (created, ready) == (want_created, want_ready):
+                return entity
+
+        self.fail("Error waiting for virtual machine. Maybe try a higher wait_timeout value?")
```

Comment:
What about `Timeout occurred while waiting on virtual machine to get to %s state`?

Reply:
done